### PR TITLE
#16 Feature: Possibility to pass many rst files at once

### PR DESCRIPTION
**Pull request by pwilczynskiclearcode**
- Added possibility to pass one or many rst files to 'rst-lint' command.
- Moved tests out of the restructuredtext_lint (reduced build package size).

### CODE REVIEW

**Comment by twolfson** [comment]
While it's great that you want to relocate the tests and reduce the build size, I prefer to bundle my tests into the repo to allow for testing packages upon download. Could we please move them back to their original positions?

File: .travis.yml [hunk 1]
```diff
@@ -2,6 +2,7 @@ language: python
 python:
   - "2.7"
   - "3.3"
+  - "3.4"
```
Comment (twolfson) [inline .travis.yml]: Please relocate this to a separate PR. I prefer PRs to focus on one thing well.

File: CHANGELOG.rst [hunk 2]
```diff
@@ -1,5 +1,8 @@
 restructuredtext-lint changelog
 ===============================
+
```
Comment (twolfson) [inline CHANGELOG.rst]: Please delete this leading newline, it breaks the existing convention for our CHANGELOG.

File: MANIFEST.in [hunk 3]
```diff
@@ -1,2 +1,2 @@
+include UNLICENSE *.rst *.py
 recursive-include restructuredtext_lint *
-include *
```
Comment (twolfson) [inline MANIFEST.in]: What was the purpose behind this change?

File: restructuredtext_lint/cli.py [hunk 4]
```diff
@@ -1,46 +1,54 @@
 import argparse
 import json
 import sys
+import locale
```
Comment (twolfson) [inline restructuredtext_lint/cli.py]: We leverage the pep8 convention of grouping and alphabetizing core modules. Please add locale in alphabetical order with the rest of the modules.

https://www.python.org/dev/peps/pep-0008/#id17

File: restructuredtext_lint/cli.py [hunk 5]
```diff
@@ -1,46 +1,54 @@
...
 
 from .lint import lint_file
 
 
-def _main(filepath, format='text', stream=sys.stdout, encoding=None):
+def _main(filepaths, format, encoding, stream=sys.stdout):
```
Comment (twolfson) [inline restructuredtext_lint/cli.py]: While I appreciate the notion of relocating defaults from `_main` to the CLI parser, it is creating noise in this PR. Please relocate those changes to a separate PR.

File: restructuredtext_lint/cli.py [hunk 6]
```diff
@@ -1,46 +1,54 @@
...
     # Read and lint the file
-    errors = lint_file(filepath, encoding=encoding)
+    file_errors = [lint_file(f, encoding=encoding) for f in filepaths]
```
Comment (twolfson) [inline restructuredtext_lint/cli.py]: Instead of performing all linting upfront and consuming memory, can we change the `for` loop to iterate over filepaths instead?

```python
error_dicts = []

for filepath in filepaths:
    errors = lint(filepath, encoding=encoding)
```

File: restructuredtext_lint/cli.py [hunk 7]
```diff
@@ -1,46 +1,54 @@
...
+    errors_dicts = []
 
-    # If there were no errors, exit gracefully
-    if not errors:
-        if format == 'json':
-            stream.write('[]')
+    for filename, errors in zip(filepaths, file_errors):
+        if not errors:
+            stream.write('INFO File {} is clean.\n'.format(filename))
```
Comment (twolfson) [inline restructuredtext_lint/cli.py]: Can we remove the 'INFO' from this log? This only seems applicable if we are consistent with it, in which case, we should be using logging.

Comment (twolfson) [inline restructuredtext_lint/cli.py]: Also, can we use the kwarg flavor for `.format`? I find that the index based version is brittle:

```python
stream.write('File "{filepath}" is clean.\n'.format(filepath=filepath))
```

Comment (twolfson) [inline restructuredtext_lint/cli.py]: Ah, I see where the 'INFO' is coming from. Never mind leave that in but I would still like the kwarg `.format`

File: restructuredtext_lint/cli.py [hunk 8]
```diff
@@ -1,46 +1,54 @@
...
         else:
-            stream.write('File was clean.\n')
-        sys.exit(0)
+            if format == 'text':
+                for error in errors:
+                    # e.g. WARNING readme.rst:12 Title underline too short.
+                    stream.write('{} {}:{} {}\n'.format(error.type, error.source, error.line, error.message))
```
Comment (twolfson) [inline restructuredtext_lint/cli.py]: Similar request for making this a kwarg flavor of `.format`:

```python
stream.write('{error.type} {error.source}:{error.line} {error.message}\n'.format(error=error))
```

File: restructuredtext_lint/cli.py [hunk 9]
```diff
@@ -1,46 +1,54 @@
...
 
 from .lint import lint_file
 
 
-def _main(filepath, format='text', stream=sys.stdout, encoding=None):
+def _main(filepaths, format, encoding, stream=sys.stdout):
     # Read and lint the file
-    errors = lint_file(filepath, encoding=encoding)
+    file_errors = [lint_file(f, encoding=encoding) for f in filepaths]
+    errors_dicts = []
```
Comment (twolfson) [inline restructuredtext_lint/cli.py]: This should be `error_dicts` as it should be a list of dictionaries. We shouldn't be pluralizing both words, only the second one.

File: restructuredtext_lint/cli.py [hunk 10]
```diff
@@ -1,46 +1,54 @@
...
-            'source': error.source,
-            'level': error.level,
-            'type': error.type,
-            'message': error.message,
-            'full_message': error.full_message,
-        } for error in errors]
-        stream.write(json.dumps(error_dicts))
+        stream.write(json.dumps(errors_dicts))
+
+    if any(file_errors):
```
Comment (twolfson) [inline restructuredtext_lint/cli.py]: When we relocate everything into a for loop, this can be simplified by adding a boolean marker inside the loop.

Comment (twolfson) [inline restructuredtext_lint/cli.py]: Example:

```python
found_errors = False
for filepath in filepaths:
    errors = lint_file(filepath, encoding=encoding)
    if errors:
      found_errors = True
```

File: restructuredtext_lint/cli.py [hunk 11]
```diff
@@ -1,46 +1,54 @@
...
-    parser.add_argument('filepath', type=str, help='File to lint')
-    parser.add_argument('--format', type=str, help='Format of output (e.g. text, json)')
-    parser.add_argument('--encoding', type=str, help='Encoding of the input file (e.g. utf-8)')
+    parser = argparse.ArgumentParser('rst-lint', description='Lint a reStructuredText files')
+    parser.add_argument('filepaths', type=str, nargs='+',
+                        help='File(s) to lint')
+    parser.add_argument('--format', '-f', type=str, default='text',
+                        help='Format of the output (e.g. text, json)')
+    parser.add_argument('--encoding', '-e', type=str, default=default_encoding,
+                        help='Encoding of the input file (e.g. {}, ascii)'.format(default_encoding))
```
Comment (twolfson) [inline restructuredtext_lint/cli.py]: One additional problem resulting from using `default_encoding` is that this will be a breaking change which means it must be a major release. Whereas the support for multiple filepaths is a non-breaking change.

File: restructuredtext_lint/cli.py [hunk 12]
```diff
@@ -1,46 +1,54 @@
...
 def main():
+    default_encoding = locale.getpreferredencoding()
+
     # Set up options and parse arguments
-    parser = argparse.ArgumentParser(description='Lint a reStructuredText file')
-    parser.add_argument('filepath', type=str, help='File to lint')
-    parser.add_argument('--format', type=str, help='Format of output (e.g. text, json)')
-    parser.add_argument('--encoding', type=str, help='Encoding of the input file (e.g. utf-8)')
+    parser = argparse.ArgumentParser('rst-lint', description='Lint a reStructuredText files')
+    parser.add_argument('filepaths', type=str, nargs='+',
```
Comment (twolfson) [inline restructuredtext_lint/cli.py]: `filepaths` reads weird under the `--help` dialogue. Can we keep it named as `filepath` for users and rename as a variable with argparse?

Currently:

```bash
$ rst-lint --help
usage: rst-lint [-h] [--format FORMAT] [--encoding ENCODING]
                filepaths [filepaths ...]

Lint a reStructuredText files

positional arguments:
  filepaths             File(s) to lint

optional arguments:
  -h, --help            show this help message and exit
  --format FORMAT, -f FORMAT
                        Format of the output (e.g. text, json)
  --encoding ENCODING, -e ENCODING
                        Encoding of the input file (e.g. UTF-8, ascii)
```

`metavar` fix:

```python
parser.add_argument('filepaths', metavar='filepath', type=str, nargs='+',
```

```bash
$ rst-lint --help
usage: rst-lint [-h] [--format FORMAT] [--encoding ENCODING]
                filepath [filepath ...]

Lint a reStructuredText files

positional arguments:
  filepath              File(s) to lint

optional arguments:
  -h, --help            show this help message and exit
  --format FORMAT, -f FORMAT
                        Format of the output (e.g. text, json)
  --encoding ENCODING, -e ENCODING
                        Encoding of the input file (e.g. UTF-8, ascii)
```

File: test.sh [hunk 13]
```diff
@@ -1,7 +1,6 @@
 #!/usr/bin/env bash
 # Exit on first error and echo commands
-set -e
-set -x
+set -ex
```
Comment (twolfson) [inline test.sh]: I prefer to keep my option setting on separate lines. Please revert this change.

File: tests/test.py [hunk 14]
```diff
@@ -23,13 +26,13 @@ def _load_file(self, filepath):
         f.close()
         return file
 
-    def _lint_file(self, *args, **kwargs):
+    def _lint_file(self, *args):
```
Comment (twolfson) [inline tests/test.py]: Instead of removing `**kwargs`, can we fix the invocation to properly pass `**kwargs`?

```
def _lint_file(self, *args, **kwargs):
    return restructuredtext_lint.lint(*args, **kwargs)
```

File: tests/test.py [hunk 15]
```diff
@@ -16,6 +18,7 @@
 
 
 class TestRestructuredtextLint(TestCase):
+
```
Comment (twolfson) [inline tests/test.py]: This extra newline breaks conventions in this repo. Please remove it.

File: tests/test.py [hunk 16]
```diff
@@ -1,11 +1,13 @@
...
 from unittest import TestCase
+from subprocess import check_output, CalledProcessError, Popen
 
 import yaml
 
 import restructuredtext_lint
 
-__dir__ = os.path.dirname(os.path.abspath(__file__))
+
+rst_dir = os.path.join(os.path.dirname(os.path.abspath(__file__)), 'test_files')
```
Comment (twolfson) [inline tests/test.py]: If we are going to support `os.path.join`, then we should clean up the slashes in the other paths as well:

```python
__dir__ = os.path.dirname(os.path.abspath(__file__))
filepath = os.path.join(__dir__, 'test_files', 'invalid.rst')
```

File: tests/test.py [hunk 17]
```diff
@@ -93,7 +96,30 @@ def test_invalid_link(self):
...
+            output = check_output('rst-lint')
+            assert "too few arguments" in output
+
+    def test_rst_lint_correct_file(self):
+        valid_file = rst_dir + '/valid.rst'
+        output = str(check_output(('rst-lint', valid_file))).splitlines()
+        assert "{} is clean".format(valid_file) in output[0]
+        assert len(output) == 1
+
+    def test_rst_lint_many_files(self):
```
Comment (twolfson) [inline tests/test.py]: For consistency with the rest of these tests, please add a docstring to each of the functions.

File: tests/test.py [hunk 18]
```diff
@@ -93,7 +96,30 @@ def test_invalid_link(self):
...
         self.assertIn('Hyperlink target "hello" is not referenced.', errors[1].message)
+
+
+class TestRestructuredtextLintCLI(TestCase):
+    """ Tests for 'rst-lint' CLI comand """
+
+    def test_rst_lint_filepaths_not_given(self):
+        with self.assertRaises(CalledProcessError):
+            output = check_output('rst-lint')
+            assert "too few arguments" in output
```
Comment (twolfson) [inline tests/test.py]: Please use `unittest's` asserts to be consistent with the rest of the tests. They provide more verbose and easier to debug messages:

```python
self.assertIn('too few arguments', output)
```

Comment (twolfson) [inline tests/test.py]: Please use single quotes over double quotes for consistency with the rest of this repo.

**Comment by twolfson** [comment]
Okay... a lot of feedback on this PR. There were a lot of undiscussed changes and inconsitencies with the repo. I am fine with discussing these changes but throwing them on a PR without talking to the maintainer is frustrating for both of us.

File: restructuredtext_lint/cli.py [hunk 19]
```diff
@@ -1,46 +1,56 @@
...
 
 from .lint import lint_file
 
 
-def _main(filepath, format='text', stream=sys.stdout, encoding=None):
-    # Read and lint the file
-    errors = lint_file(filepath, encoding=encoding)
+def _main(filepaths, format='text', stream=sys.stdout, encoding=None):
+    # Read and lint the files
+    errors = []
```
Comment (pwilczynskiclearcode) [inline restructuredtext_lint/cli.py]: Still `errors` is not memory optimal. I would need to resign from `json.dumps` in favour of printing json partials on my own or to find some json iterative library that could do the trick.

**Comment by pwilczynskiclearcode** [comment]
Removed all changes considered unrelated to the problem being fixed.

File: restructuredtext_lint/cli.py [hunk 20]
```diff
@@ -5,42 +5,47 @@
...
+        sys.exit(0)
 
 
 def main():
     # Set up options and parse arguments
-    parser = argparse.ArgumentParser(description='Lint a reStructuredText file')
-    parser.add_argument('filepath', type=str, help='File to lint')
-    parser.add_argument('--format', type=str, help='Format of output (e.g. text, json)')
-    parser.add_argument('--encoding', type=str, help='Encoding of the input file (e.g. utf-8)')
+    parser = argparse.ArgumentParser('rst-lint', description='Lint reStructuredText files')
```
Comment (twolfson) [inline restructuredtext_lint/cli.py]: We present the script under multiple names (`restructuredtext-lint` and `rst-lint`) so naming it one or the other isn't too practical =/

**Comment by twolfson** [comment]
Thanks for the quick reply, fixes, and respecting my opinion =)

It looks like my comments regarding the test changes (not the renaming but the contents of the file) were lost though. I am going to re-post them =/

File: restructuredtext_lint/cli.py [hunk 21]
```diff
@@ -5,42 +5,47 @@
...
+    parser.add_argument('filepaths', metavar='filepath', nargs='+', type=str, help='File to lint')
+    parser.add_argument('--format', default='text', type=str, help='Format of the output (text or json)')
```
Comment (twolfson) [inline restructuredtext_lint/cli.py]: Can we move this back to `e.g.`? This is providing examples and not enforcing that we will always be mutually exclusive (e.g. we could support `xml`, `yaml`).

File: restructuredtext_lint/test/test.py [hunk 22]
```diff
@@ -23,13 +26,13 @@ def _load_file(self, filepath):
         f.close()
         return file
 
-    def _lint_file(self, *args, **kwargs):
+    def _lint_file(self, *args):
```
Comment (twolfson) [inline restructuredtext_lint/test/test.py]: Instead of removing `**kwargs`, can we fix the invocation to properly pass `**kwargs`?

```
def _lint_file(self, *args, **kwargs):
    return restructuredtext_lint.lint(*args, **kwargs)
```

File: restructuredtext_lint/test/test.py [hunk 23]
```diff
@@ -16,6 +18,7 @@
 
 
 class TestRestructuredtextLint(TestCase):
+
```
Comment (twolfson) [inline restructuredtext_lint/test/test.py]: This extra newline breaks conventions in this repo. Please remove it.

File: restructuredtext_lint/test/test.py [hunk 24]
```diff
@@ -1,11 +1,13 @@
...
+from subprocess import check_output, CalledProcessError
 from unittest import TestCase
 
 import yaml
 
 import restructuredtext_lint
 
-__dir__ = os.path.dirname(os.path.abspath(__file__))
+
+rst_dir = os.path.join(os.path.dirname(os.path.abspath(__file__)), 'test_files')
```
Comment (twolfson) [inline restructuredtext_lint/test/test.py]: If we are going to support `os.path.join`, then we should clean up the slashes in the other paths as well:

```python
__dir__ = os.path.dirname(os.path.abspath(__file__))
# ...
filepath = os.path.join(__dir__, 'test_files', 'invalid.rst')
```

File: restructuredtext_lint/test/test.py [hunk 25]
```diff
@@ -93,7 +96,30 @@ def test_invalid_link(self):
...
+        filepath = rst_dir + '/invalid_link.rst'
         errors = restructuredtext_lint.lint_file(filepath)
         self.assertIn('Anonymous hyperlink mismatch: 1 references but 0 targets.', errors[0].message)
         self.assertIn('Hyperlink target "hello" is not referenced.', errors[1].message)
+
+
+class TestRestructuredtextLintCLI(TestCase):
+    """ Tests for 'rst-lint' CLI comand """
+
+    def test_rst_lint_filepaths_not_given(self):
```
Comment (twolfson) [inline restructuredtext_lint/test/test.py]: For consistency with the rest of these tests, please add a docstring to each of the functions.

File: restructuredtext_lint/test/test.py [hunk 26]
```diff
@@ -93,7 +96,30 @@ def test_invalid_link(self):
...
+        with self.assertRaises(CalledProcessError):
+            output = check_output('rst-lint')
+            assert "too few arguments" in output
```
Comment (twolfson) [inline restructuredtext_lint/test/test.py]: Please use `unittest's` asserts to be consistent with the rest of the tests. They provide more verbose and easier to debug messages:

```python
self.assertIn('too few arguments', output)
```

Comment (twolfson) [inline restructuredtext_lint/test/test.py]: Please use single quotes over double quotes for consistency with the rest of this repo.

File: restructuredtext_lint/cli.py (hunk 21 above)
Comment (pwilczynskiclearcode) [inline restructuredtext_lint/cli.py]: We could, but we don't, so it's kind of misleading. You give false hopes that other formats are supported.

**Comment by pwilczynskiclearcode** [comment]
Just an observation:
There's a lot of conventions that are hard to notice since they aren't written down and tests/lints don't cover them. It would be extremely helpful if I started coding with the list of your required conventions.

File: restructuredtext_lint/cli.py (hunk 21 above)
Comment (twolfson) [inline restructuredtext_lint/cli.py]: Yea but they are easily added/supported. This is a personal preference and I would prefer to use `e.g.`as that's my typical dictation.

**Comment by twolfson** [comment]
Yea, sorry about that. Once this PR is landed, I will add in some linting/enforcement.

**Comment by twolfson** [comment]
Regarding JSON being inefficient memory wise, I totally agree and we should consider moving to a streaming JSON output in the future.

File: restructuredtext_lint/test/test.py [hunk 27]
```diff
@@ -93,7 +95,33 @@ def test_invalid_link(self):
...
+        """The `rst-lint` command prints out 'X is clean' if rst file is correct."""
+        valid_file = os.path.join(__dir__, 'test_files', 'valid.rst')
+        output = str(check_output(('rst-lint', valid_file))).splitlines()
+        self.assertIn('{filepath} is clean'.format(filepath=valid_file), output[0])
+        self.assertEqual(len(output), 1)
+
+    def test_rst_lint_many_files(self):
+        """The `rst-lint` command accepts many rst file paths and prints respective information for each of them."""
+        with self.assertRaises(CalledProcessError) as e:
+            check_output('rst-lint ' + os.path.join(__dir__, 'test_files', '*rst'), shell=True)
```
Comment (twolfson) [inline restructuredtext_lint/test/test.py]: Please don't use concatenation. If someone has a space in their pathname, then the test will fail. Instead, pass these as separate arguments:

```python
check_output(('rst-lint ', os.path.join(__dir__, 'test_files', '*rst')), shell=True)
```

File: restructuredtext_lint/test/test.py [hunk 28]
```diff
@@ -93,7 +95,33 @@ def test_invalid_link(self):
...
         self.assertIn('Hyperlink target "hello" is not referenced.', errors[1].message)
+
+
+class TestRestructuredtextLintCLI(TestCase):
+    """ Tests for 'rst-lint' CLI comand """
+
+    def test_rst_lint_filepaths_not_given(self):
+        """The `rst-lint` command is available and prints error if no filepath was given."""
+        with self.assertRaises(CalledProcessError):
+            output = check_output('rst-lint')
```
Comment (twolfson) [inline restructuredtext_lint/test/test.py]: Is `rst-lint` always going to resolve to the working version? Maybe we should do something like:

```python
check_output(('python', os.path.join(__dir__, '..', 'cli.py'))
```

Comment (twolfson) [inline restructuredtext_lint/test/test.py]: If we choose to do this, then maybe make it a constant next to `__dir__`:

```python
rst_lint_path = os.path.join(__dir__, '..', 'cli.py')
# ...
check_output(('python', rst_lint_path))
```

**Comment by pwilczynskiclearcode** [comment]
> Regarding JSON being inefficient memory wise, I totally agree and we should consider moving to a streaming JSON output in the future.

I couldn't find any library that would do it. Python `json.JSONEncoder` actually can output chunks of JSON but still it requires a dumped object to be in memory. 
I find that the only possibility is to write JSON chunks on our own: `'['`, `'{"level": 4, "source": "invalid_line_mismatch.rst", "full_message": "Title overline & underline mismatch.\n\n=====\nb\n====", "line": 1, "type": "SEVERE", "message": "Title overline & underline mismatch."}'`, ..., `']'`

**Comment by pwilczynskiclearcode** [comment]
> Regarding JSON being inefficient memory wise, I totally agree and we should consider moving to a streaming JSON output in the future.

I couldn't find any library that would do it. Python `json.JSONEncoder` actually can output chunks of JSON but still it requires a dumped object to be in memory. 
I find that the only possibility is to write JSON chunks on our own: `'['`, `'{"level": 4, "source": "invalid_line_mismatch.rst", "full_message": "Title overline & underline mismatch.\n\n=====\nb\n====", "line": 1, "type": "SEVERE", "message": "Title overline & underline mismatch."}'`, ..., `']'`

**Comment by twolfson** [comment]
JSONStreamWriter seems to fit our use case. However, we should not worry about that in this PR:

https://github.com/lysdexia/python-json-stream-writer

File: restructuredtext_lint/test/test.py [hunk 29]
```diff
@@ -1,11 +1,17 @@
...
 
 import yaml
 
 import restructuredtext_lint
 
+
 __dir__ = os.path.dirname(os.path.abspath(__file__))
+valid_rst = os.path.join(__dir__, 'test_files', 'valid.rst')
+invalid_rst = os.path.join(__dir__, 'test_files', 'invalid.rst')
+rst_lint_path = os.path.join(__dir__, os.pardir, 'cli.py')
```
Comment (twolfson) [inline restructuredtext_lint/test/test.py]: TIL about `os.pardir`. From the docs it looks like it's `..` in all cases but I guess they like future proofing.

https://docs.python.org/2/library/os.html#os.pardir

File: restructuredtext_lint/test/test.py [hunk 30]
```diff
@@ -93,7 +97,36 @@ def test_invalid_link(self):
...
+        """The `rst-lint` command prints out 'X is clean' if rst file is correct."""
+        raw_output = check_output((sys.executable, rst_lint_path, valid_rst))
+        output = str(raw_output).splitlines()
+        self.assertIn('{filepath} is clean'.format(filepath=valid_rst), output[0])
+        self.assertEqual(len(output), 1)
+
+    def test_rst_lint_many_files(self):
+        """The `rst-lint` command accepts many rst file paths and prints respective information for each of them."""
+        with self.assertRaises(CalledProcessError) as e:
+            check_output((sys.executable, rst_lint_path, valid_rst, invalid_rst))
```
Comment (twolfson) [inline restructuredtext_lint/test/test.py]: It would be nice to provide an example of what this executes but I can add that on while landing

**Comment by twolfson** [comment]
Everything looks good. Landing this PR shortly. Thanks for all your work and quick responses =)

**Comment by twolfson** [comment]
There was some noise in the test runner due to not capturing `stderr` from `check_output`. I am patching this before landing.

```
$ nosetests --nocapture $*
.......usage: cli.py [-h] [--format FORMAT] [--encoding ENCODING]
              filepath [filepath ...]
cli.py: error: too few arguments
.
----------------------------------------------------------------------
Ran 8 tests in 0.136s

OK
```

**Comment by twolfson** [comment]
This has been merged/released in `0.12.0`. Thanks again for all your work =)

**Comment by pwilczynskiclearcode** [comment]
:+1:
